src/compiler/propagation/type_propagator.h

```diff
@@ -84,41 +86,45 @@ class TypePropagator {
 
   Map<uint8*, Set<TypeVariable*>> sites_;
 
-  std::unordered_map<uint8*, std::vector<MethodTemplate*>> templates_;
+  std::unordered_map<uint32, MethodTemplate*> methods_;
+  std::unordered_map<uint32, BlockTemplate*> blocks_;
+
   std::unordered_map<int, TypeVariable*> globals_;
   std::unordered_map<uint8*, TypeVariable*> outers_;
   std::unordered_map<unsigned, std::unordered_map<int, TypeVariable*>> fields_;
   std::vector<MethodTemplate*> enqueued_;
 
   void call_method(MethodTemplate* caller, TypeScope* scope, uint8* site, Method target, std::vector<ConcreteType>& arguments);
 
-  MethodTemplate* find(Method target, std::vector<ConcreteType> arguments);
-  MethodTemplate* instantiate(Method method, std::vector<ConcreteType> arguments);
+  MethodTemplate* find_method(Method target, std::vector<ConcreteType> arguments);
+  BlockTemplate* find_block(MethodTemplate* origin, Method method, int level);
 };
 
 class MethodTemplate {
  public:
-  MethodTemplate(TypePropagator* propagator, Method method, std::vector<ConcreteType> arguments)
-      : propagator_(propagator)
+  MethodTemplate(MethodTemplate* link, TypePropagator* propagator, Method method, std::vector<ConcreteType> arguments)
+      : link_(link)
+      , propagator_(propagator)
       , method_(method)
       , arguments_(arguments)
       , result_(propagator->words_per_type()) {}
 
+  ~MethodTemplate() {
+    delete link_;
+  }
+
+  MethodTemplate* link() const { return link_; }
```

Comment:
I find `link` to be a bad name. Makes me think of an action (the verb "link").
Since this is a linked list, I would just go with `next`.

I would also add a comment on the class, explaining that it has a field for a linked-list. You could mention where it is used for, but I would be ok with just mentioning the fact that it is a linked-list element.